src/codeeditor/editor.ts

```diff
@@ -0,0 +1,637 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable
+} from 'phosphor/lib/core/disposable';
+
+import {
+  find
+} from 'phosphor/lib/algorithm/searching';
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  ISignal, clearSignalData, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  IChangedArgs
+} from '../common/interfaces';
+
+import {
+  ObservableVector
+} from '../common/observablevector';
+
+import {
+  JSONObject
+} from 'phosphor/lib/algorithm/json';
+
+
+/**
+ * A namespace for code editors.
+ */
+export
+namespace CodeEditor {
+  /**
+   * A zero-based position in the editor.
+   */
+  export
+  interface IPosition {
+    /**
+     * The cursor line number.
+     */
+    line: number;
+
+    /**
+     * The cursor column number.
+     */
+    column: number;
+  }
+
+  /**
+   * An interface describing editor state coordinates.
+   */
+  export
+  interface ICoords {
+    /**
+     * The left coordinate value.
+     */
+    readonly left: number;
+
+    /**
+     * The right coordinate value.
+     */
+    readonly right: number;
+
+    /**
+     * The top coordinate value.
+     */
+    readonly top: number;
+
+    /**
+     * The bottom coordinate value.
+     */
+    readonly bottom: number;
+  }
+
+  /**
+   * A text selection.
+   */
+  export
+  interface ITextSelection {
+    /**
+     * The index to the first character in the current selection.
+     *
+     * #### Notes
+     * If this position is greater than [end] then the selection is considered
+     * to be backward.
+     */
+    start: number;
+
+    /**
+     * The index to the last character in the current selection.
+     *
+     * #### Notes
+     * If this position is less than [start] then the selection is considered
+     * to be backward.
+     */
+    end: number;
+
+    /**
+     * The uuid of the text selection owner.
+     */
+    uuid: string;
+
+    /**
+     * A class name added to the text selection.
+     */
+    className?: string;
+  }
+
+  /**
+   * An editor model.
+   */
+  export
+  interface IModel extends IDisposable {
+    /**
+     * A signal emitted when the value changes.
+     */
+    valueChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * A signal emitted when a property changes.
+     */
+    mimeTypeChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * The text stored in the model.
+     */
+    value: string;  // TODO: this should be an iobservablestring.
+
+    /**
+     * A mime type of the model.
+     */
+    mimeType: string;
+
+    /**
+     * Get the number of lines in the model.
+     */
+    readonly lineCount: number;
+
+    /**
+     * The currently selected code.
+     *
+     * @returns A read-only copy of the text selections.
+     */
+    readonly selections: ObservableVector<ITextSelection>;
```

Comment:
@blink1073 I've prototyped [another interface](https://github.com/R-Brain/jupyterlab/pull/3/files#diff-e5ea874541b5876668fdda0d2fc94e89R122) to manage selections. An idea that a selection owner is a single source of changes and users should use them instead of directly manipulate model to avoid bidirectional syncing, but they still can observe the model.

There is [ISelectionOwner](https://github.com/R-Brain/jupyterlab/pull/3/files#diff-e5ea874541b5876668fdda0d2fc94e89R391) interface and [IEditor](https://github.com/R-Brain/jupyterlab/pull/3/files#diff-e5ea874541b5876668fdda0d2fc94e89R441) implements it to provide access to a cursor and selections. There can be other implementations of `ISelectionOwner`, e.g. remote users who work with the same file.

I've update text area example to fulfill this contract:
- the editor directly access the text area to manage a cursor position and selections: https://github.com/R-Brain/jupyterlab/pull/3/files#diff-e5ea874541b5876668fdda0d2fc94e89R824
- the editor propagates selection changes to a model: https://github.com/R-Brain/jupyterlab/pull/3/files#diff-e5ea874541b5876668fdda0d2fc94e89R879
- the editor observes selection changes from other owners and display them, it does not need to update own selections, since nobody else can change it: https://github.com/R-Brain/jupyterlab/pull/3/files#diff-e5ea874541b5876668fdda0d2fc94e89R864